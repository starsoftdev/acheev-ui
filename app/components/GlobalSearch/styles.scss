@import '~styles/settings/settings';
.globalSearch {
  position: relative;

  @include breakpoint(small only) {
    background-color: $lighter-gray;
    margin: 0 (-$rem-sm) (-$rem-sm) (-$rem-sm);
    border-top: 1px solid $light-gray;
    height: rem-calc(40);

    &--open {
      height: rem-calc(340);
    }
  }
  &__dropDown {
    position: absolute;
    top: rem-calc(45);
    width: 100%;
    max-height: rem-calc(300);
    overflow-y: scroll;

    padding: 0;

    background-color: $white;

    border: 1px solid $light-gray;
    border-radius: rem-calc(4);

    z-index: 2;

    @include breakpoint(small only) {
      border-radius: 0;
      border-left: none;
      border-right: none;
      position: absolute;
      top: rem-calc(40);
    }
  }

  &__subSection {
    padding: 0 $rem-lg;

    &:first-child {
      border-top: none;
    }
  }

  &__category {
    display: block;
    color: $secondary;

    font-size: $fs-tn;
    font-weight: bolder;
    text-transform: uppercase;
    text-decoration: none;

    padding: $rem-md 0 $rem-sm 0;

    border-top: 1px solid $light-gray;

    @at-root {
      .globalSearch__subSection:first-child & {
        border-top: none;
      }
    }
  }

  &__linkItem {
    display: block;
    color: $darker-gray;
    padding: $rem-sm $rem-lg;
    margin-left: -$rem-lg;
    margin-right: -$rem-lg;
    font-size: $fs-mn;
    text-decoration: none;
    text-transform: none;
    &:hover {
      color: $white;
      background-color: $secondary;
    }
  }

  &__noData {
    padding: $rem-lg;
    text-align: center;
  }
}
