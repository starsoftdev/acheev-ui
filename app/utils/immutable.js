// @flow

import { Iterable } from 'immutable';
import type { Map } from 'immutable';
import { get as _get } from 'lodash-es';

export const get = (
  collection: ?Map<*, *> | Object,
  path: string | number,
  defaultValue: any
) => {
  if (!collection) {
    return defaultValue;
  } else if (Iterable.isIterable(collection)) {
    return collection.get(path, defaultValue);
  }
  return _get(collection, path, defaultValue);
};

export const getIn = (
  collection: ?Map<*, *> | Object,
  path: Array<string | number>,
  defaultValue: any
) => {
  if (!collection) {
    return defaultValue;
  } else if (Iterable.isIterable(collection)) {
    return collection.getIn(path, defaultValue);
  }
  return _get(collection, path, defaultValue);
};
