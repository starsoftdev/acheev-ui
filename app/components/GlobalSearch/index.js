// @flow

import React, { Component } from 'react';
import Link from 'components/Link';
import cx from 'classnames';
import { fromJS } from 'immutable';
import type { Map } from 'immutable';

import SearchField from 'components/SearchField';
import Preloader from 'components/Preloader';

import './styles.scss';

type Props = {
  requestGlobalSearch: Function,
  filter: Map<*, *>,
  data: Map<string, Object>,
  isLoading?: boolean,
};

type State = {
  dropDownVisible: boolean,
  searchValue: string,
};

class GlobalSearch extends Component<Props, State> {
  static defaultProps = {
    filter: fromJS({}),
  };

  state = {
    dropDownVisible: false,
    searchValue: '',
  };

  componentWillMount() {
    const { filter } = this.props;
    this.setState({
      searchValue: filter.getIn(['query', 'q'], ''),
    });
    this.typingTimeout = 0;
  }

  componentDidMount() {
    window.addEventListener('click', this.closeDropDown, false);
  }

  componentWillReceiveProps(newProps: Props) {
    this.setState({
      searchValue: newProps.filter.getIn(['query', 'q'], ''),
    });
  }

  componentWillUnmount() {
    window.removeEventListener('click', this.closeDropDown, false);
  }

  onInputChange = (value: string) => {
    if (this.typingTimeout) {
      clearTimeout(this.typingTimeout);
    }
    const timeoutId = setTimeout(() => {
      this.searchProduct(this.state.searchValue);
    }, 1000);
    this.setState({
      searchValue: value,
      dropDownVisible: value !== '',
    });
    this.typingTimeout = timeoutId;
  };

  onInputFocus = () => {
    const { searchValue } = this.state;
    this.setState({ dropDownVisible: searchValue !== '' });
  };

  onClickDropdown = e => {
    e.stopPropagation();
  };

  typingTimeout: number;

  closeDropDown = () => {
    this.setState({ dropDownVisible: false });
  };

  searchProduct = (value: string) => {
    this.props.requestGlobalSearch(['query', 'q'], value);
  };

  render() {
    const { dropDownVisible } = this.state;
    const { data, isLoading } = this.props;

    let listContent = (
      <div className="globalSearch__noData">No results found.</div>
    );

    if (data && data.size > 0) {
      listContent = data.entrySeq().map(([key, category]) => (
        <div key={key} className="globalSearch__subSection">
          <Link
            className="globalSearch__category"
            to={`/${key}`}
            onClick={this.closeDropDown}
          >
            {key}
          </Link>
          {category &&
            category.entrySeq().map(([index, value]) => (
              <Link
                key={index}
                className="globalSearch__linkItem"
                to={value.get('link')}
                onClick={this.closeDropDown}
              >
                {value.getIn(['item', 'name'])}
              </Link>
            ))}
        </div>
      ));
    }
    return (
      <div
        className={cx('globalSearch', {
          'globalSearch--open': dropDownVisible === true,
        })}
      >
        <div className="row align-middle nm">
          <div className="column np">
            <SearchField
              className="global"
              onChange={this.onInputChange}
              onFocus={this.onInputFocus}
              defaultValue={this.state.searchValue}
              onClick={this.onClickDropdown}
              placeholder="What skills are you looking for?"
            />
          </div>
        </div>
        {dropDownVisible === true && (
          <div
            className="globalSearch__dropDown"
            onClick={this.onClickDropdown}
            role="button"
          >
            <div className="row column">
              {isLoading ? (
                <Preloader height={300} />
              ) : (
                <div>{listContent}</div>
              )}
            </div>
          </div>
        )}
      </div>
    );
  }
}

export default GlobalSearch;
