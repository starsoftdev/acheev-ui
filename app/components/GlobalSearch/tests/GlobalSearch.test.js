import React from 'react';
import { shallow } from 'enzyme';
import { create } from 'react-test-renderer';
import { fromJS } from 'immutable';

import GlobalSearch from 'components/GlobalSearch';

const data = fromJS({
  advice: [
    {
      createdOn: '2017-10-04T06:52:22.557Z',
      score: 0.5833333333333334,
      item: {
        name: 'Hey canada post messed up, which made me wonder...',
        ref: '57ff92928b999fa13a17b32f',
      },
      __v: 0,
      _id: '59d48526517e554b342dc63b',
      updatedOn: '2017-10-04T06:52:22.557Z',
      link:
        'https://lift.co/forum/laws-and-regulations/hey-canada-post-messed-up-which-made-me-wonder',
      id: '59d48526517e554b342dc63b',
      category: 'advice',
    },
    {
      createdOn: '2017-10-04T06:52:22.236Z',
      score: 0.5294117647058824,
      item: {
        name:
          'Hey everyone, just thought you should know that when you write a review, your user name &amp; member bio information is released to the public via google',
        ref: '56b93dc1569880a977f72502',
      },
      __v: 0,
      _id: '59d48526517e554b342dc511',
      updatedOn: '2017-10-04T06:52:22.236Z',
      link:
        'https://lift.co/forum/laws-and-regulations/hey-everyone-just-thought-you-should-know-that-when-you-write-a-review-your-user-name-amp-member-bio-information-is-released-to-the-public-via-google',
      id: '59d48526517e554b342dc511',
      category: 'advice',
    },
  ],
});
const filter = fromJS({
  query: {
    q: 'hey',
  },
});

describe('<GlobalSearch />', () => {
  it("shouldn't throw error if data is not specified", () => {
    try {
      shallow(<GlobalSearch />);
    } catch (error) {
      expect(error).toBeUndefined();
    }
  });

  it('submitFunction should be fired when `GlobalSearch__submitButton` is clicked', () => {
    const spy = jest.fn();
    const renderedComponent = shallow(
      <GlobalSearch requestGlobalSearch={spy} />
    );
    renderedComponent.find('SearchField').simulate('change');
    setTimeout(() => expect(spy).toHaveBeenCalled());
  });

  it('submitFunction should be fired when `GlobalSearch__submitButton` is clicked', () => {
    const renderedComponent = shallow(<GlobalSearch isLoading />);
    renderedComponent.find('SearchField').simulate('change');
    expect(renderedComponent.find('Preloader').length).toEqual(1);
  });

  it('renders `data` and `filter` if provided', () => {
    const component = create(<GlobalSearch data={data} filter={filter} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
