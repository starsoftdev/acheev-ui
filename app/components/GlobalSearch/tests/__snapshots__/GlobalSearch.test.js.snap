// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GlobalSearch /> renders \`data\` and \`filter\` if provided 1`] = `
<div
  className="globalSearch"
>
  <div
    className="row align-middle nm"
  >
    <div
      className="column np"
    >
      <div
        className="searchField global"
      >
        <input
          className="searchField__input dark"
          defaultValue="hey"
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyPress={[Function]}
          placeholder="Search for strains, oils, producers ..."
          type="text"
        />
        <button
          className="button searchField__button white"
          onClick={[Function]}
        >
          <div
            className="row align-middle align-center"
          >
            <div
              className="shrink column"
            >
              <svg
                className="searchField__icon"
                height={14}
                width={14}
              >
                <use
                  xlinkHref="#undefined"
                />
              </svg>
            </div>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;
